python/src/aac/validate/_validate_definition.py

```diff
@@ -0,0 +1,123 @@
+) -> list[ValidatorResult]:
+    """Traverse the definition validate it."""
+
+    validator_results = []
+    ancestor_definitions = get_definition_ancestry(definition, context)
+
+    # For each ancestor definition, get validation definitions, collapse the 2d list to 1d, and return the list of validations.
+    applicable_ancestor_validation_dicts = list(flatten(map(_get_validation_entries, ancestor_definitions)))
+
+    # Ancestor validators apply to the whole definition structure
+    for validation_dict in applicable_ancestor_validation_dicts:
+        validation_name = validation_dict.get("name")
+        validator_plugin = _get_validator_plugin_by_name(validation_name, validator_plugins)
+
+        if validator_plugin:
+            validator_results.append(_apply_validator(validation_dict, validator_plugin))
+
+    # Traverse through the definition and validate its fields and sub-structures.
+    validator_results.extend(_validate_definition_substructure(definition, validator_plugins, context))
+
+    return validator_results
+
+
+def _validate_definition_substructure(
+    definition: ParsedDefinition, validator_plugins: list[ValidatorPlugin], context: ActiveContext
+) -> list[ValidatorResult]:
+    """Return a list of validator results from recursively validating the populated fields of the definition."""
+    validator_results = []
+
+    def _apply_validation(content_to_validate: dict, validators: list[ValidatorPlugin]) -> list[ValidatorResult]:
+        validator_results.extend(
+            [_apply_validator(content_to_validate, context, validator) for validator in validators]
+        )
+
+    def _recursively_apply_field_validations(definition_to_validate: ParsedDefinition) -> None:
+        """Recurse through the fields and identify validations to apply."""
+        definition_fields_dict = get_definition_fields_and_types(definition_to_validate, context.definitions)
+
+        for field_name, field_definition in definition_fields_dict.items():
+            fields_to_validate = search_definition(definition_to_validate, [definition_to_validate.get_root_key(), field_name])
+
+            # Skip empty fields, primitive types, self-defining endless loops
+            if (
+                not fields_to_validate
+                or not hasattr(field_definition, "name")
+                or definition_to_validate.name == field_definition.name
+            ):
+                continue
+
+            field_type_validations = _get_validation_entries(field_definition)
+            applicable_validators = [
+                _get_validator_plugin_by_name(validation.get("name"), validator_plugins)
+                for validation in field_type_validations
+            ]
+
+            # If the field is not an array type
+            if isinstance(fields_to_validate, dict):
+                _apply_validation(fields_to_validate, applicable_validators)
+                # Need to recurse here for embedded definitions
```

Comment:
Is this effectively a TODO comment? If so, could we make that into a story (if it's not already done) so we don't forget to do that?